### DISCUSSION TITLE
Reporting diagnostics for syntax nodes without diagnostic analyzer context

### DISCUSSION THREAD

**Discussion by veenadange** [dotnet/roslyn#68053] [Q&A] [answered]
Hi Team,

I need to analyze all invocations for a particular method and check the value of it's argument. The method argument can be a variable and it's declaration can be in a different project of the solution. Based on whether the value of the declaration is valid/invalid, a diagnostic should be reported. 

I am trying to implement a custom 'Diagnostic analyzer' for this, which works fine for declarations of argument variable in same project but not for cross project declaration (which gives empty DeclaringSyntaxReferences for argument Symbol). As far as I came across, that will not be possible as analyzers run over compilation and there is no context of project/solution/workspace.

As an alternative if I separate out this analyzer functionality as a stand-alone tool/custom command, will it be possible to report diagnostics for the targeted project/solution on which this runs? Without analyzer I can't get any context object to report diagnostics on.

Or is there any other elegant way to implement such use cases?

Thank you!

**Comment by CyrusNajmabadi** [comment]
You'll need to make a separate standalone tool if you want cross project analysis. As your own tool you create whatever sort of diagnostics you want and report them however you want.

**Reply by veenadange** [reply]
Thanks for the reply @CyrusNajmabadi. Yes, I can generate a file or so as a report but I wanted to check is there any way it can reflect as a diagnostic like a sort of warning on particular statement of that document if it's open in IDE? Just the same way as a diagnostic analyzer would have reported it.

**Reply by CyrusNajmabadi** [reply, marked as answer]
>  is there any way it can reflect as a diagnostic like a sort of warning on particular statement of that document if it's open in IDE? Just the same way as a diagnostic analyzer would have reported it.

Using roslyn: no. 
Using the VS apis: yes.

In Roslyn, the only way to support this is by being a DiagnosticAnalyzer, and that is subject to the limitations outlined above.  

In VS though, you can make whatever squiggles/task-list items you want.  It's not easy though, and you'll have to do a lot of work maintaining that code.  Sorry!

**Reply by veenadange** [reply]
Yes, VS Api seems to be the only way! I got my answer to main query, thanks for that! 

Just one more thing I want to ask that whether it is possible to programmatically build multiple solutions having projects with variety of .net versions accurately using MsBuildWorkspace? It seems tidious as paths to Msbuild dlls needs to be supplied while registering workspace and even if we register all possible paths, how a project can be compiled using the right version?
